text/3430-unix-socket-ancillary-data-v2.md

```diff
@@ -0,0 +1,802 @@
+        ancillary_data: &mut AncillaryData<'_>,
+        options: SendOptions,
+    ) -> io::Result<usize>;
+}
+
+trait SendMessageTo {
+    type SocketAddr;
+
+    fn send_message_to(
+        &self,
+        addr: &Self::SocketAddr,
+        bufs: &[IoSlice<'_>],
+        ancillary_data: &mut AncillaryData<'_>,
+        options: SendOptions,
+    ) -> io::Result<usize>;
+}
+```
+
+The `SendMessage` trait will be implemented for:
+- `std::net::TcpStream`
+- `std::net::UdpSocket`
+- `std::os::unix::net::UnixDatagram`
+- `std::os::unix::net::UnixStream`
+
+The `SendMessageTo` trait will be implemented for:
+- `std::net::UdpSocket`
+- `std::os::unix::net::UnixDatagram`
+
+These traits aren't sealed, so they can be implemented by third-party libraries.
+
+### `SendOptions`
+
+The `std::os::unix::net::SendOptions` type wraps flags passed to the Unix
+sockets API functions [`send`] and [`sendmsg`]. It has inherent methods for
+options defined by the POSIX standard.
+
+[`send`]: https://pubs.opengroup.org/onlinepubs/9699919799/functions/send.html
+[`sendmsg`]: https://pubs.opengroup.org/onlinepubs/9699919799/functions/sendmsg.html
+
+```rust
+struct SendOptions;
+
+impl Default for SendOptions;
+
+impl SendOptions {
+    fn new() -> SendOptions;
+
+    fn as_send_flags(&self) -> c_int;
+
+    fn custom_flags(&mut self, flags: c_int) -> &mut Self;
+
+    // MSG_EOR
+    fn end_of_record(&mut self, end_of_record: bool) -> &mut Self;
+
+    // MSG_OOB
+    fn out_of_band(&mut self, out_of_band: bool) -> &mut Self;
+
+    // MSG_NOSIGNAL
+    fn no_signal(&mut self, no_signal: bool) -> &mut Self;
+}
+```
+
+### `MessageSender`
+
+The `MessageSender` helper struct provides an ergonomic API around the
+`SendMessage` / `SendMessageTo` traits.
+
+```rust
+struct MessageSender<S, 'a>;
+
+impl<S, 'a> MessageSender<S, 'a> {
+    fn new(
+        socket: &'a S,
+        buf: &'a [u8],
+    ) -> MessageSender<S, 'a>;
+
+    fn new_vectored(
+        socket: &'a S,
+        bufs: &'a [IoSlice<'a>],
+    ) -> MessageSender<S, 'a>;
+
+    fn ancillary_data(
+        &mut self,
+        ancillary_data: &'a mut AncillaryData<'a>,
+    ) -> &mut MessageSender<S, 'a>;
+
+    fn options(
+        &mut self,
+        options: SendOptions,
+    ) -> &mut MessageSender<S, 'a>;
+}
+
+impl<S: SendMessage> MessageSender<S, '_> {
+    fn send(&mut self) -> io::Result<usize>;
+}
+
+impl<S: SendMessageTo> MessageSender<S, '_> {
+    fn send_to(&mut self, addr: &S::SocketAddr) -> io::Result<usize>;
+}
+```
+
+## Receiving messages
+
+### The `RecvMessage` and `RecvMessageFrom` traits
+
+These traits are implemented for socket types that implement `recvmsg`. They
+are equivalent to the current `recv_vectored_with_ancillary` and
+`recv_vectored_with_ancillary_from` functions, but also accept `RecvOptions`
+for per-recv flags, and aren't specific to `AF_UNIX` sockets.
+
+```rust
+trait RecvMessage {
+    fn recv_message(
+        &self,
+        bufs: &mut [IoSliceMut<'_>],
```

Comment:
> The use of `&[IoSlice]` / `&mut [IoSliceMut]` reflects the underlying `sendmsg` / `recvmsg` API, which operates on an array of `struct iovec`. The existing API (`recv_vectored_with_ancillary`, etc) has the same structure.

I'd like the `recvmsg` API to accept a `BorrowedCursor` instead.

`std::io::Read` is also heading the same direction to avoid zero-initialization required for reading data.

Reply:
@jmillikin I still think it's better to use `BorrowedCursor`, or some API `&mut [MaybeUninitIoSliceMut]` to accept uninitialized buffer.

Reply:
You'd need the `read_buf` feature to add support for vectored I/O first. The tracking issue describes this as "Vectorized APIs (`read_buf_vectored`, `ReadBufs`)."

With a vectorized version in place, the appropriate trait might look like this:

```rust
trait RecvMessageBuf {
    fn recv_message(
        &self,
        bufs: &mut std::io::ReadBufs,
        ancillary_data: &mut AncillaryData<'_>,
        options: RecvOptions,
    ) -> io::Result<(usize, RecvResult)>;
}
```

Note that you'd need a way to convert from `ReadBufs` to `struct iovec *`, similar to the type layout guarantees of `IoSliceMut`.

Implementing that code is beyond the scope of this RFC. You need to follow up on https://github.com/rust-lang/rust/issues/78485 to ask for vectored I/O support.

Reply:
Thanks!

I guess `RecvMessageBufs` can upgrade to `ReadBuf` once https://github.com/rust-lang/rust/issues/78485 is redady, just like `std::io::Read`.